### DISCUSSION TITLE
Difficulty with Yolo Classification Performance with Object Detection Model

### DISCUSSION THREAD

**Discussion by nkinnaird** [ultralytics/ultralytics#17115] [Q&A]
Hi there, I am using Ultralytics Yolo code to train a YoloV8 object detection model on frames from videos. These frames contain surgical instruments in the fore view of the frame which I am attempting to localize and classify. The surgical instruments can currently be one of 3 classes, and look similar except for the metal parts of the tool. Ie. the object is ~80% the same from one class to another, and the last ~20% is different. 

The bounding box prediction part of the model is working quite well with predicted bounding boxes encasing pretty much all tools encountered in the frame. However the classification performance for the tools themselves is not doing so well, with subpar performance. Ie. the model is confusing the classes of the tools in various cases, and in a lot of cases where it's obvious to a human what the class is. I've tried so many things to get this to work, including:

```
	- More augmentations, less augmentations
		○ Rotation, scale, translation, color augs, various albumentations, mixup values, cutout, mosaic
	- Random weights, pretrained
	- Smaller models, larger models (all yolo, nano small medium large)
	- Yolov8, yolov11
	- Different cls weight in loss function, bumping it up to higher values
	- Adding more data, different data
	- Additional passes through data labeling to try and fix any unclear cases, mislabels, etc.
	- Reducing frames from many-frame videos to try and reduce oversampling in some cases
	- Filters on box sizes, box positions
	- Single class detection model and secondary classification model
	- Different learning rates, values, strategies, optimizers
	- Longer training, shorter training
        - Adjusting nms parameters, clustering parameters
```

I don't have a ton of labeled images, but I _think_ it should be enough for what I'm doing:

`Train set: Total number of jpg files: 4463 (2279 class 0, 623 class 1, 2351 class 2, 2467 class 3, 648 class 4)`
`Validation set: Total number of jpg files: 1810 (1409 class 0, 212 class 1, 811 class 2, 853 class 3, 232 class 4)`

(There are two other classes, class 1 and class 4 which are other miscellaneous things, I mostly care about classes 0, 2, and 3 which are the surgical tools). 

I've spent weeks on this now with only minor improvements at various stages for things I've tried, but always worse performance than I need. As far as I can tell the model is learning how to classify the tools based on looking at the backgrounds and positions of the tools rather than the metal parts of the tools themselves. My experience is that augmentations help with this kind of thing but attempts at this have not helped.

Has anyone dealt with this before? Is there anything I haven't tried that comes to mind? Anything that I might be doing wrong? I would appreciate any and all ideas, even if it's just a throwaway thought, thanks!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @nkinnaird, thank you for reaching out to us about your experience with Ultralytics 🚀! It sounds like you've been working hard to optimize your YOLOv8 model for classifying surgical instruments. 

If this is a 🐛 Bug Report, it would be very helpful if you could share a minimum reproducible example to help us investigate the issue further.

For classification performance issues like the one you're experiencing, I recommend double-checking that the dataset is well-balanced for the classes of interest and that the annotations accurately reflect what you want the model to learn. Additionally, it might be beneficial to experiment with different data augmentation strategies, ensuring that these help the model focus on distinguishing the crucial metal parts of the tools. 

You've tried quite a comprehensive list of strategies already, but sometimes a slight refinement in the loss function or further data curation can provide the necessary improvements. 

An Ultralytics engineer will also assist shortly to provide more insights and suggestions. In the meantime, feel free to join our community chat on Discord for real-time assistance, interactions, and support from fellow users 🎧.

**Comment by glenn-jocher** [comment]
It sounds like you've tried a comprehensive set of strategies. Given the subtle differences between classes, consider using a higher-resolution input to capture more detail or employing a model with a focus on fine-grained classification. Additionally, exploring feature visualization might help understand what the model is focusing on. For more insights, you can refer to our common issues guide at https://docs.ultralytics.com/guides/yolo-common-issues/.

**Reply by nkinnaird** [reply]
When you say a model for fine-grained classification what do you mean? I'd like to just use a single shot model like Yolo if possible since it gives me the bounding boxes and the classes at the same time, though I did try a secondary classification model to no success. (I didn't spend a ton of time on it so maybe more work there would be fruitful, but then improving and maintaining 2 models becomes painful...)

From feature visualization with https://github.com/rigvedrs/YOLO-V8-CAM (EigenCAM / GradCAM) it appears as though the model is paying too much attention the backgrounds, and not the tools themselves. Not sure what I can do with that information other than I now know the model isn't paying attention to what I want it to. (For one of the tools it does pay attention to the metal parts, and for that tool the performance is better, though still not as high as I'd like.)

**Reply by glenn-jocher** [reply]
@nkinnaird consider using a higher-resolution input to capture more detail in the metal parts. You might also try adjusting the model's architecture to focus more on fine-grained features. For further insights, you can explore different attention mechanisms or specialized layers that enhance feature extraction.
